### ISSUE TITLE
TrackedStorage (flagged component storage)

### ISSUE THREAD

**Issue by Aceeri** [slide-rs/specs#89]
Having a `TrackedStorage` that would add/remove flags depending on if the component was modified would be a good bonus to performance in a lot of cases. For instance, if you only want to recalculate something once the component is modified.

One problem is if the user uses a mutable storage and iterates instead of joining then it will flag all of the components. This could be solved by comparing the two results on mutable iterations.

**Comment by Aceeri** [comment]
Adding associated types to the `UnprotectedStorage` may make this easier and less painful to use.
e.g.
```rust
trait UnprotectedStorage<'a, T> {
    type Target: 'a;
    type TargetMut: 'a;
    ...
    unsafe fn get(&self, id: Index) -> Target;
    unsafe fn get(&mut self, id: Index) -> TargetMut;
    ...
}
```

This would allow for the storages to return components with a wrapper type for various purposes, in this case it would track if the component was modified through `Deref/DerefMut`.

Current issues with this means adding a lifetime to the UnprotectedStorage bubbles up:
`UnprotectedStorage<'a, T> + Component<'a> + MaskedStorage<'a, T>`

It isn't an issue until you reach the typemap of `TypeId -> Storage`, since the type id requires a `'static` bound and we require a variable/non-'static bound.

Issue for non-'static TypeId: https://github.com/rust-lang/rust/issues/41875

**Comment by torkleyy** [comment]
From Gitter:

We decided to have a `FlaggedStorage` and a `ChangedStorage`.
The first one just flags components which were iterated over mutably (or fetched using `get_mut`), while the latter one does an equality check.

**Comment by torkleyy** [comment]
As commented above, this issue has been split into two parts. `FlaggedStorage` was already implemented by #93. Closing this now and creating a new issue for the `ChangedStorage`.
